### PR TITLE
perf(laguna): align MoE router and verify paths

### PR DESCRIPTION
**Pull request by davide221**
## What This PR Achieves

This is the Laguna autoregressive parity patch with llama.cpp. It keeps PR #455 scoped to the AR target path and removes the later speculative/DDTree experiments from the PR diff.

Main changes:
- Aligns Laguna MoE routing with llama.cpp's sigmoid -> top-k -> get_rows pattern so ggml-cuda can use the top-k MoE fusion path.
- Adds a small shared `moe_router_graph` helper for the fused router node order.
- Auto-enables Laguna head-major KV layout for the plain AR decode path when it is safe to do so.
- Adds head-major KV cache/snapshot/attention views and 256-aligned KV capacity for the Laguna FA path.
- Keeps Laguna KV cache type selection on the existing `DFLASH_KV_TYPE`, `DFLASH_KV_K_TYPE`, and `DFLASH_KV_V_TYPE` controls.
- Bumps the llama.cpp submodule to the companion ggml/CUDA change used by the parity path.

Cleaned scope:
- `5 files changed`
- `261 insertions`
- `90 deletions`
- one clean commit authored by `davide221` with `Codex` as co-author

## Validation

Checked locally:

```bash
git diff --check 26ca0dcb6a198628b6cab2b0a2ffbb901baaa564 b8a9f46643674a662524fffa402d54f1593ef4a2
```

Built on the Laguna CUDA bench box:

```bash
cmake -B /home/lucebox/lucebox-hub-pr455-clean-code/server/build-pr455-clean-cuda \
  -S /home/lucebox/lucebox-hub-pr455-clean-code/server \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_CUDA_ARCHITECTURES=86

cmake --build /home/lucebox/lucebox-hub-pr455-clean-code/server/build-pr455-clean-cuda \
  --target bench_laguna_generate dflash_server -j 8
```

The clean build produced both `bench_laguna_generate` and `dflash_server`. BSA was disabled in this copied verification tree because only the llama.cpp submodule was synced; that does not affect this Laguna AR compile check.

## Not Included

- DDTree behavior changes.
- GPU speculative accept / verify plumbing.
- Speculative fallback/gating/profiling experiments.
- Domino-head experiments.
- SWA-window experiments.
- The later experimental MoE down+combine kernel, which was not byte-identical.

Companion ggml/CUDA PR:
https://github.com/Luce-Org/llama.cpp-dflash-ggml/pull/23

### CODE REVIEW

**Comment by davide221** [comment]
Final PR update pushed: `130e26d perf(laguna): avoid greedy logits readback`.

Patch:
- Laguna greedy decode now uses in-graph GPU argmax and skips full logits D2H. Sampling still reads logits.
- Escape hatch: `DFLASH_LAGUNA_GPU_ARGMAX=0` forces the old full-logits CPU argmax path.
- `bench_laguna_generate` has `DFLASH_LAGUNA_BENCH_NO_LOGITS=1` for apples-to-apples llama-bench `tg` semantics.

Same RTX 3090, synced dependency commit `8b11267c4`:
- standalone llama-bench `tg512`: `187.05 +/- 0.47 tok/s`; `pp128`: `1334.49 +/- 3.96 tok/s`
- Lucebox core no-logits decode (`prompt_N=1`, `n_gen=512`): `191.3 tok/s`
- Lucebox real greedy decode (`prompt_N=128`, `n_gen=512`, lm_head + GPU argmax): `181.1 tok/s`
- Lucebox OpenAI server request (`75 in / 256 out`): decode `180.6 tok/s`, overall `163.6 tok/s`

Validation:
- `cmake --build server/build-pr23sync-cuda --target bench_laguna_generate dflash_server -j 8`
- fallback smoke: `DFLASH_LAGUNA_GPU_ARGMAX=0 ... 32 32` generated the same synthetic id stream.

**Comment by davide221** [comment]
Cleaned and pushed the Laguna greedy decode graph reuse patch in `be324bd`.

Validation on the Lucebox RTX 3090:

- Clean build: `bench_laguna_generate`, `dflash_server`, `test_dflash` passed.
- Token generation check: old PR head vs patched build generated byte-identical ID streams for `bench_laguna_generate 128 384` (`GENERATED_IDS_MATCH=1`).
- Core Laguna AR decode: old `181.0 tok/s`, patched `187.4 tok/s` on 384 tokens; patched `186.8 tok/s` on 512 tokens.
- HTTP server HumanEval streaming, no draft/DDTree, `n=10`, `n_gen=256`: mean decode `188.04 tok/s`, wall `167.67 tok/s`.

The removed dependency-level `mul_mat->argmax` experiment was intentionally not included because it fired correctly but gave no measurable speedup.
